Reduce unsafeness in WebCore/Modules/WebGPU

annevk opened the pull request:
#### d098479be5daad1a6944784c420e95e3947ebc11
<pre>
Reduce unsafeness in WebCore/Modules/WebGPU
<a href="https://bugs.webkit.org/show_bug.cgi?id=291367">https://bugs.webkit.org/show_bug.cgi?id=291367</a>

Reviewed by Mike Wyrzykowski.

* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromImageBuffer):
(WebCore::getImageBytesFromVideoFrame):
(WebCore::imageBytesForSource):
(WebCore::isOriginClean):
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::getMappedRange):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
(WebCore::WebGPU::DeviceImpl::createShaderModule):
(WebCore::WebGPU::convertToBacking):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.cpp:
(WebCore::WebGPU::GPUImpl::createPresentationContext):
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBuffer.h:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::getMappedRange):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.h:

Canonical link: <a href="https://commits.webkit.org/293568@main">https://commits.webkit.org/293568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d40447cf1f451c3ac3fc0346026a756fbb8df498

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99305 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18954 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9203 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104436 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49906 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101346 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27388 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75575 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32679 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102312 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14625 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89660 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55935 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14419 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7641 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49266 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84344 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106793 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26419 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19264 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84534 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26781 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85862 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84046 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28712 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6404 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20151 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16150 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26359 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31553 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26179 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29492 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27746 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/WebGPU/GPUBuffer.h:
```diff
@@ -66,6 +66,7 @@ class GPUBuffer : public RefCountedAndCanMakeWeakPtr<GPUBuffer> {
 
     WebGPU::Buffer& backing() { return m_backing; }
     const WebGPU::Buffer& backing() const { return m_backing; }
+    Ref<const WebGPU::Buffer> protectedBacking() const { return m_backing; }
```
I think we can just mark the data member as `const` and not need a protected getter.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -409,7 +409,7 @@ static void getImageBytesFromVideoFrame(WebGPU::Queue& backing, const RefPtr<Vid
 #endif
 
-static void imageBytesForSource(WebGPU::Queue& backing, const auto& sourceDescriptor, const auto& destination, bool& needsYFlip, bool& needsPremultipliedAlpha, ImageDataCallback&& callback)
+static void imageBytesForSource(WebGPU::Queue& backing, const auto& sourceDescriptor, const auto& destination, bool& needsYFlip, bool& needsPremultipliedAlpha, NOESCAPE ImageDataCallback&& callback)
```
nit: if callback is indeed called synchronously, I think it'd be nicer to use `NOESCAPE const ImageDataCallback&` since we're not transferring ownership or intending to store the callback somewhere.

mwyrzykowski replied:
Yes, its always synchronous, there is a release assertion on line 1007 ensuring this `RELEASE_ASSERT(callbackScopeIsSafe);`

Changing to a `const&` sounds good to me 👍

cdumez commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
```diff
@@ -251,8 +251,8 @@ RefPtr<PipelineLayout> DeviceImpl::createPipelineLayout(const PipelineLayoutDesc
     if (descriptor.bindGroupLayouts) {
-        backingBindGroupLayouts = descriptor.bindGroupLayouts->map([&convertToBackingContext = m_convertToBackingContext.get()](auto bindGroupLayout) {
-            return convertToBackingContext.convertToBacking(bindGroupLayout.get());
+        backingBindGroupLayouts = descriptor.bindGroupLayouts->map([convertToBackingContext = m_convertToBackingContext](auto bindGroupLayout) {
```
Vector::map() uses `NOESCAPE`, I'm surprised we needed to make any change here?

cdumez commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
```diff
@@ -387,8 +387,8 @@ static auto convertToBacking(const ComputePipelineDescriptor& descriptor, Conver
 {
-    return convertToBacking(descriptor, m_convertToBackingContext, [backing = m_backing, &convertToBackingContext = m_convertToBackingContext.get()](const WGPUComputePipelineDescriptor& backingDescriptor) {
-        return ComputePipelineImpl::create(adoptWebGPU(wgpuDeviceCreateComputePipeline(backing.get(), &backingDescriptor)), convertToBackingContext);
+    return convertToBacking(descriptor, m_convertToBackingContext, [backing = m_backing, convertToBackingContext = m_convertToBackingContext](const WGPUComputePipelineDescriptor& backingDescriptor) {
```
Can we mark the `convertToBacking()` lambda as `NOESCAPE const T&` instead?

annevk replied:
I couldn't figure this one out. I left this unchanged for now.

cdumez commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
```diff
@@ -585,8 +585,8 @@ static auto convertToBacking(const RenderPipelineDescriptor& descriptor, Convert
 {
-    return convertToBacking(descriptor, m_convertToBackingContext, [backing = m_backing.copyRef(), &convertToBackingContext = m_convertToBackingContext.get()](const WGPURenderPipelineDescriptor& backingDescriptor) {
-        return RenderPipelineImpl::create(adoptWebGPU(wgpuDeviceCreateRenderPipeline(backing.get(), &backingDescriptor)), convertToBackingContext);
+    return convertToBacking(descriptor, m_convertToBackingContext, [backing = m_backing.copyRef(), convertToBackingContext = m_convertToBackingContext](const WGPURenderPipelineDescriptor& backingDescriptor) {
```
ditto.

cdumez commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
```diff
@@ -599,8 +599,8 @@ static void createComputePipelineAsyncCallback(WGPUCreatePipelineAsyncStatus sta
 {
-    convertToBacking(descriptor, m_convertToBackingContext, [backing = m_backing.copyRef(), &convertToBackingContext = m_convertToBackingContext.get(), callback = WTFMove(callback)](const WGPUComputePipelineDescriptor& backingDescriptor) mutable {
-        auto blockPtr = makeBlockPtr([convertToBackingContext = Ref { convertToBackingContext }, callback = WTFMove(callback)](WGPUCreatePipelineAsyncStatus status, WGPUComputePipeline pipeline, String&& message) mutable {
+    convertToBacking(descriptor, m_convertToBackingContext, [backing = m_backing.copyRef(), convertToBackingContext = m_convertToBackingContext, callback = WTFMove(callback)](const WGPUComputePipelineDescriptor& backingDescriptor) mutable {
```
ditto.

cdumez commented on Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBuffer.h:
```diff
@@ -50,7 +50,7 @@ class Buffer : public RefCountedAndCanMakeWeakPtr<Buffer> {
 
     virtual void mapAsync(MapModeFlags, Size64 offset, std::optional<Size64>, CompletionHandler<void(bool)>&&) = 0;
-    virtual void getMappedRange(Size64 offset, std::optional<Size64>, Function<void(std::span<uint8_t>)>&&) = 0;
+    virtual void getMappedRange(Size64 offset, std::optional<Size64>, NOESCAPE Function<void(std::span<uint8_t>)>&&) = 0;
```
Ditto, I think it is nicer to use `NOESCAPE const Function<void(std::span<uint8_t>)>&` if the lambda gets called synchronously and not stored. There is no reason to require a r-value reference.

mwyrzykowski replied:
Agreed, it's used with an IPC call but currently via a sendSync call so const& sounds good

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -405,11 +405,11 @@ static void getImageBytesFromVideoFrame(WebGPU::Queue& backing, const RefPtr<Vid
     vImagePermuteChannels_ARGB8888(&bgra, &bgra, permuteMap, kvImageNoFlags);
 
-    return callback(byteSpan.first(sizeInBytes), width, height);
+    callback(byteSpan.first(sizeInBytes), width, height);
```
This line seems unnecessarily changed unless I am misunderstanding, but also it is fine without the return

annevk replied:
Yeah this was drive-by cleanup as the method is declared as returning void.

mwyrzykowski replied:
it was like that intentionally for consistency with the other return statements but I am also fine with the change :)

annevk replied:
I see, I can undo it before landing. Though note that `getImageBytesFromImageBuffer` doesn't have a return statement either. But maybe that's the inconsistency from your perspective?

mwyrzykowski replied:
Oh yes it doesn't really matter so much.

The thing it prevents is someone changes the return type of `ImageDataCallback` but then fails to change the return type of the calling function (`getImageBytesFromVideoFrame`) which the `return someFunctionReturningVoid` style prevents.

In this case it doesn't matter due to the earlier above return statements which would catch the same issue. And certainly there are other ways to prevent these mistakes too (`WARN_UNUSED_RETURN`). It would be nice if `WARN_UNUSED_RETURN` was the default and then no need to have this pattern which is arguably confusing.

mwyrzykowski reviewed (approved):
Thank you!

annevk committed:
Reduce unsafeness in WebCore/Modules/WebGPU
https://bugs.webkit.org/show_bug.cgi?id=291367

Reviewed by Mike Wyrzykowski.

* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromImageBuffer):
(WebCore::getImageBytesFromVideoFrame):
(WebCore::imageBytesForSource):
(WebCore::isOriginClean):
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::getMappedRange):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
(WebCore::WebGPU::DeviceImpl::createShaderModule):
(WebCore::WebGPU::convertToBacking):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.cpp:
(WebCore::WebGPU::GPUImpl::createPresentationContext):
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBuffer.h:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::getMappedRange):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.h:

Canonical link: https://commits.webkit.org/293568@main
